src/equations/shallow_water_two_layer_1d.jl

```diff
@@ -0,0 +1,586 @@
+
+Further details are available in Theorem 1 of the paper:
+- Niklas Wintermeyer, Andrew R. Winters, Gregor J. Gassner and David A. Kopriva (2017)
+  An entropy stable nodal discontinuous Galerkin method for the two dimensional
+  shallow water equations on unstructured curvilinear meshes with discontinuous bathymetry
+  [DOI: 10.1016/j.jcp.2017.03.036](https://doi.org/10.1016/j.jcp.2017.03.036)
+"""
+#TODO: Change name after paper is finished
+@inline function flux_wintermeyer_etal(u_ll, u_rr,
+                                       orientation::Integer,
+                                       equations::ShallowWaterTwoLayerEquations1D)
+  # Unpack left and right state
+  h1_ll, h1_v1_ll, h2_ll, h2_v2_ll, _ = u_ll
+  h1_rr, h1_v1_rr, h2_rr, h2_v2_rr, _ = u_rr
+
+  # Get the velocities on either side
+  v1_ll, v2_ll = velocity(u_ll, equations)
+  v1_rr, v2_rr = velocity(u_rr, equations)
+
+  # Average each factor of products in flux
+  v1_avg = 0.5 * (v1_ll + v1_rr)
+  v2_avg = 0.5 * (v2_ll + v2_rr)
+  p1_avg = 0.5 * equations.gravity * h1_ll * h1_rr
+  p2_avg = 0.5 * equations.gravity * h2_ll * h2_rr
+
+  # Calculate fluxes
+  f1 = 0.5 * (h1_v1_ll + h1_v1_rr)
+  f2 = f1 * v1_avg + p1_avg
+  f3 = 0.5 * (h2_v2_ll + h2_v2_rr)
+  f4 = f3 * v2_avg + p2_avg
+
+  return SVector(f1, f2, f3, f4, zero(eltype(u_ll)))
+end
+
+
+"""
+    flux_es(u_ll, u_rr, orientation,
+            equations::ShallowWaterTwoLayerEquations1D)
+
+Entropy stable surface flux for the two-layer shallow water equations. Uses the entropy 
+conservative flux_fjordholm_etal and adds a Lax-Friedrichs type dissipation dependent on the jump 
+of entropy variables.
+
+Further details are available in the paper:
+- Ulrik Skre Fjordholm (2012)
+  Energy conservative and stable schemes for the two-layer shallow water equations.
+  [DOI: 10.1142/9789814417099_0039](https://doi.org/10.1142/9789814417099_0039)
+It should be noted that the equations are ordered differently and the
+designation of the upper and lower layer has been changed which leads to a slightly different
+formulation.
+"""
+@inline function flux_es(u_ll, u_rr,
+                         orientation::Integer,
+                         equations::ShallowWaterTwoLayerEquations1D)
+  # Compute entropy conservative flux but without the bottom topography
+  f_ec = flux_fjordholm_etal(u_ll, u_rr,
+                              orientation,
+                              equations)[1:4]
+
+  # Get maximum signal velocity
+  λ = max_abs_speed_naive(u_ll, u_rr, orientation, equations)
+
+  # Get entropy variables but without the bottom topography
+  q_rr = cons2entropy(u_rr,equations)[1:4]
+  q_ll = cons2entropy(u_ll,equations)[1:4]
+
+  # Average values from left and right
+  u_avg = (u_ll + u_rr) / 2
+
+  # Introduce variables for better readability
+  rho1 = equations.rho1
+  rho2 = equations.rho2
+  g    = equations.gravity
+  drho = rho1 - rho2
+
+  # Entropy Jacobian matrix
+  H = Array{Float64,2}(undef,4,4)
```

Comment:
This allocates and will be slow. Moreover, it does not allow AD. You could check whether something like
```suggestion
  H = MArray{Tuple{4, 4}, eltype(u_ll), 2}(undef)(undef)
```
works better.

Reply:
I changed the array type to MArray in commit 0e45df1d684b4edbe9ba03e7c5dbd90a934bfbb7